include/rosparam_shortcuts/rosparam_shortcuts.h

```diff
@@ -106,6 +110,246 @@ bool convertDoublesToEigen(const std::vector<double>& values, Eigen::Isometry3d&
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name, default_value);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+template <typename ParameterT>
+void NodeParameters::declare_and_get(const std::string& name, const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+void NodeParameters::register_ros_callback_reconfigure() {
+	on_set_callback_handle_ = node_->add_on_set_parameters_callback(
+	    [this](const std::vector<rclcpp::Parameter>& parameters) { return set_parameters_callback(parameters); });
+}
+
+void NodeParameters::register_update_callback(const std::function<void()>& callback) const {
+	changed_notify_callback_.push_back(callback);
+}
+
+bool NodeParameters::get(const std::string& name, rclcpp::Parameter& parameter) const {
+	auto internal_param = parameters_.find(name);
+	if (internal_param != parameters_.end()) {
+		parameter = parameters_.at(name);
+		return true;
+	}
+
+	return false;
+}
+
+rclcpp::Parameter NodeParameters::get(const std::string& name) const {
+	auto internal_param = parameters_.find(name);
+	if (internal_param != parameters_.end()) {
+		return parameters_.at(name);
+	} else {
+		RCLCPP_ERROR_STREAM(logger_, "Parameter: " << name + " not found.");
+	}
+
+	return rclcpp::Parameter();
+}
+
+rcl_interfaces::msg::SetParametersResult
+NodeParameters::set_parameters_callback(const std::vector<rclcpp::Parameter>& parameters) {
+	const std::lock_guard<std::mutex> guard(mutex_);
+	auto result = rcl_interfaces::msg::SetParametersResult();
+	result.successful = true;
+
+	// First validate
+	for (const auto& parameter : parameters) {
+		const auto internal_param = parameters_.find(parameter.get_name());
+		const auto search = validate_.find(parameter.get_name());
+
+		// Only checks if parameter was declared
+		if (search != validate_.end()) {
+			// will fail if parameter is not in internalized
+			if (parameter.get_type_name() != internal_param->second.get_type_name()) {
+				result.successful = false;
+				result.reason = "Parameter: " + parameter.get_name() + " type does not match internal type.";
+				return result;
+			}
+
+			// Parameter callback
+			result = search->second(parameter);
+			if (!result.successful) {
+				// Handle automatic parameter set failure
+				return result;
+			}
+		}
+
+		else {
+			result.successful = false;
+			result.reason = "Parameter: " + parameter.get_name() + " was not declared.";
+			return result;
+		}
+	}
+
+	// Set parameters
+	for (const auto& parameter : parameters) {
+		RCLCPP_INFO_STREAM(logger_, "setParametersCallback - "
```

Comment:
In the next version (g-turtle) of ros I'd recommend making this DEBUG instead of INFO because we can finally configure the logger level per-namespace.